Substantial rewrite of the Cookbook guide

brian-rose opened the pull request:
<!--
Project Pythia has automated review requests. If you are not ready for reviews on this contribution, please
open this Pull Request as a "Draft" by clicking the dropdown on the green "Create Pull Request" button in the
lower right corner here.
-->
This PR get the Cookbook guide up to date and adds substantial detail on working with the Pythia infrastructure.

Closes #433

brian-rose commented:
Here's a direct link to the preview of the new guide, since it's a bit hard to get to from the regular preview link: https://projectpythia.org/_preview/435/cookbook-guide.html

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,250 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
+- `.github/workflows/publish-book.yaml`
+- `.github/workflows/trigger-book-build.yaml`
+
+In each of these files, in the field called `environment_name:`, replace  `cookbook-dev` with the name your used in your `environment.yml` file (probably `<your-cookbook-name>-dev`). Commit these changes.
```
```suggestion
In each of these files, in the field called `environment_name:`, replace  `cookbook-dev` with the name you used in your `environment.yml` file (probably `<your-cookbook-name>-dev`). Commit these changes.
```

r-ford commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,250 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
 
-## Add your cookbook to the Cookbooks Gallery!
+Any time you push new content to the `main` branch of your Cookbook repo, actions are triggered that do the following:
+- create an up-to-date build environment using your `environment.yml` file
+- execute all the notebooks in that environment
+- package the book into a styled website
+- check all the links in the website to make sure they are valid
+- deploy the styled website so it's visible to the world
```
```suggestion
- Create an up-to-date build environment using your `environment.yml` file
- Execute all the notebooks in that environment
- Package the book into a styled website
- Check all the links in the website to make sure they are valid
- Deploy the styled website so it's visible to the world
```

r-ford reviewed (approved):
I think all the changes here make sense.

This document is pretty large now, and I think it could be confusing for someone seeing it for the first time. I guess this is really a theme issue, but I can barely differentiate the heading levels, and there is no sidebar on the right with the section outline.

brian-rose commented:
> I think all the changes here make sense.
> 
> This document is pretty large now, and I think it could be confusing for someone seeing it for the first time. I guess this is really a theme issue, but I can barely differentiate the heading levels, and there is no sidebar on the right with the section outline.

I had the same thought, I wished the level 2 and level 3 headings were easier to distinguish. I can try adding lines to separate the major sections.

Is the lack of a right sidebar a known theme issue? I see a button "toggle secondary sidebar" but it is non-functional.

brian-rose commented:
Unfortunately horizontal lines don't help much because they are nearly invisible.

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,261 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
-1. On the [Cookbook Template repository](https://github.com/ProjectPythia/cookbook-template), click "Use this template&rarr;Create a new repository"
+1. On the [Cookbook Template repository](https://github.com/ProjectPythia/cookbook-template), click "Use this template &rarr; Create a new repository"
 1. Give your repository a descriptive name followed by `-cookbook` (e.g., `hydrology-cookbook`, `hpc-cookbook`, `cesm-cookbook`) and a description
 1. Choose "Include all branches" and create the repository. Your browser will be directed to the newly created repository under your GitHub account
```
Small nit, but the option to "include all branches" occurs above the repository name selection

brian-rose replied:
Resolved: https://projectpythia.org/_preview/435/cookbook-guide.html#use-the-template

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,261 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
-1. [Clone the repository](https://foundations.projectpythia.org/foundations/github/github-cloning-forking.html) in your local workspace
-1. Within `environment.yml`, change `name` from `cookbook-dev` to `<your-cookbook-name>-dev` (e.g. `cesm-cookbook-dev`) and add all required libraries and other dependencies under `dependencies:`. Commit the changes
-1. Create the Conda environment with `conda env create -f environment.yml`. If it crashes, try running `conda config --set channel_priority strict`
+1. Under Settings &rarr; Pages, ensure that GitHub Pages is enabled. If it is not, change the Branch from "None" to "gh-pages/(root)" and click "Save"
```
```suggestion
1. Under Settings &rarr; Pages, ensure that GitHub Pages is enabled by checking that `Branch` is set to "gh-pages", and the folder set to "gh-pages/(root)". If it is not, change the Branch from "None" to "gh-pages/(root)" and click "Save"
```

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,261 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
-1. Within `environment.yml`, change `name` from `cookbook-dev` to `<your-cookbook-name>-dev` (e.g. `cesm-cookbook-dev`) and add all required libraries and other dependencies under `dependencies:`. Commit the changes
-1. Create the Conda environment with `conda env create -f environment.yml`. If it crashes, try running `conda config --set channel_priority strict`
+1. Under Settings &rarr; Pages, ensure that GitHub Pages is enabled. If it is not, change the Branch from "None" to "gh-pages/(root)" and click "Save"
+1. Under Settings &rarr; Actions &rarr; General, allow Github Actions to **push** to the repository <img width="901" alt="Screenshot 2023-01-13 at 3 12 47 PM" src="https://user-images.githubusercontent.com/26660300/212428991-cd0ae2f0-73ca-40d8-b983-f122359463aa.png">
```
The word "push" does not appear the page. The included image suggest that "Read and write permissions", under Workflow permissions", should be selected. Is this an additional step, or what is meant by allowing "push"?

brian-rose replied:
I'll clarify the text. We mean that "Read and write permissions" should be selected. No additional steps.

brian-rose replied:
Resolved: https://projectpythia.org/_preview/435/cookbook-guide.html#use-the-template

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,261 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
+1. Accept the GitHub invitation. _Welcome to the team!_
+1. Once you are a member of the organization, navigate to the Settings of your cookbook repository and scroll down to the **Danger Zone**.
+1. Click "Transfer".
+1. Select or type "ProjectPythia", confirm, and transfer.
```
There is another step following this one that asks which teams to give access to. Options are nobody, core, education, infrastructure, outreach, and "find a team". What is the guidance here?

brian-rose replied:
Hmm I guess the guidance should be to select "core". I think that includes all the active members of both "education" and "infrastructure". We haven't been very careful or consistent about how we use those teams.

brian-rose replied:
I selected "core" when I transferred a [new cookbook repo](https://github.com/ProjectPythia/radiative-feedback-cookbook) the other day, which is what got me started on this guide revision.

I forgot to include this guidance in the text.

brian-rose replied:
Resolved: https://projectpythia.org/_preview/435/cookbook-guide.html#steps-to-transfer-the-repository

clyne reviewed (commented):
Review blocked until repository transfer issue is resolved

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,250 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
+
+### Build your Cookbook locally
+
+You should be able to build your new Cookbook on your local machine by doing
```
```suggestion
You should be able to build your new Cookbook on your local machine by running this command from the root of the repository:
```

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,250 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
+1. Click "Transfer".
+1. Select or type "ProjectPythia", confirm, and transfer.
+
+Once you have successfully transfered the repository, you'll most likely want to make a [personal fork and a local clone of the repository](https://foundations.projectpythia.org/foundations/github/github-cloning-forking.html) so that you can continue to develop and collaborate on the Cookbook via the [forking workflow](https://foundations.projectpythia.org/foundations/github/github-workflows.html ...
```
After creating a personal fork of the repo I subsequently learned that GitHub had disabled all of the workflows and I had to re-enable them.

brian-rose replied:
> After creating a personal fork of the repo I subsequently learned that GitHub had disabled all of the workflows and I had to re-enable them.

This is expected! Normally we only want the workflows to operate on the upstream fork hosted in the ProjectPythia org. The typical and supported way to preview the changes that you make in your personal fork is to open a pull request on the upstream fork.

I can add this as a clarification in the guide.

There *may* be use cases where it makes sense to enable the workflows in the personal fork, but honestly I have never needed this for own cookbook development work.

clyne replied:
Gotcha. Clarifying would help.

brian-rose replied:
Clarification is now in the note block at the end of [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book).

clyne reviewed (commented):
Thanks for doing this, @brian-rose. It was badly needed

Following the instructions I was able to create a [test cookbook](https://github.com/ProjectPythia/hello-pythia-cookbook).  However, I did receive an error in one of the GitHub Actions as noted in my in-line comments. 

Also, a gh-pages website was never published, either in my personal repo or in the ProjectPythia repo for the cookbook. Not sure how to fix that.

brian-rose commented:
> Also, a gh-pages website was never published, either in my personal repo or in the ProjectPythia repo for the cookbook. Not sure how to fix that.

It's here: https://projectpythia.org/hello-pythia-cookbook/README.html

brian-rose commented:
> Following the instructions I was able to create a [test cookbook](https://github.com/ProjectPythia/hello-pythia-cookbook). However, I did receive an error in one of the GitHub Actions as noted in my in-line comments.

From what I can see [here](https://github.com/ProjectPythia/hello-pythia-cookbook/actions), all the actions are working as expected in the upstream fork.

brian-rose commented:
@clyne thanks for the detailed read-through! I think I have addressed all your comments, and I added a bit more detail on making Pull Requests.

In retrospect, I think it might be better to separate this guide into two pieces:
- contributing to Cookbooks
- making new Cookbooks

But that could be left as an open issue for the future if we're reasonably happy with what's here.

clyne commented:
> > Also, a gh-pages website was never published, either in my personal repo or in the ProjectPythia repo for the cookbook. Not sure how to fix that.
> 
> It's here: https://projectpythia.org/hello-pythia-cookbook/README.html

Sorry, should have been more clear. The gh-pages site is not linked from underneath the "About" heading on the repo [website](https://github.com/ProjectPythia/hello-pythia-cookbook). Shouldn't this happen automatically?

clyne commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,256 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
+### About the publishing pipeline
 
-## Make a Zenodo release of your Cookbook
+Any time you push new content to the `main` branch of your Cookbook repo, actions are triggered that do the following:
```
So "your Cookbook repo" is a little misleading here? The actions are only triggered on the cookbook that is in the ProjectPythia organization, not your fork of the repo, correct?

brian-rose replied:
I added a note block at the bottom of [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book) to try to clarify this. 

We can use more precise language in the sentence you highlighted. I didn't want it to get too convoluted. Strictly speaking, it's the `main` branch of the `upstream` repository, not your personal fork. Unless you *haven't* transferred ownership of your repo, it which case it would be the `main` branch of your personal repository (which is not a fork and thus does not have an upstream equivalent).

brian-rose replied:
Any suggested language for this?

clyne replied:
Nope. That looks good.

clyne reviewed (commented):
Just a couple more minor in-line comments.

brian-rose commented:
> > > Also, a gh-pages website was never published, either in my personal repo or in the ProjectPythia repo for the cookbook. Not sure how to fix that.
> > 
> > 
> > It's here: https://projectpythia.org/hello-pythia-cookbook/README.html
> 
> Sorry, should have been more clear. The gh-pages site is not linked from underneath the "About" heading on the repo [website](https://github.com/ProjectPythia/hello-pythia-cookbook). Shouldn't this happen automatically?

No, it does not happen automatically. Maybe you missed my suggestion for this in [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book) under "Here's a handy trick..."

brian-rose commented:
> > > > > Also, a gh-pages website was never published, either in my personal repo or in the ProjectPythia repo for the cookbook. Not sure how to fix that.
> > > > 
> > > > 
> > > > It's here: https://projectpythia.org/hello-pythia-cookbook/README.html
> > > 
> > > 
> > > Sorry, should have been more clear. The gh-pages site is not linked from underneath the "About" heading on the repo [website](https://github.com/ProjectPythia/hello-pythia-cookbook). Shouldn't this happen automatically?
> > 
> > 
> > No, it does not happen automatically. Maybe you missed my suggestion for this in [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book) under "Here's a handy trick..."
> 
> No, I didn't miss it, but it explains the rabbit hole that I went down. The "About" settings are different for my personal repo, and do not include the option to "Use your GitHub Pages website". I probably got confused about which repo I was looking at. Here's a screenshot for what I see on my personal repo:
> 
> <img alt="About" width="637" src="https://private-user-images.githubusercontent.com/6777736/337478475-66c0ee96-959e-4c0b-8614-1e03a1c15f05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3MjA1NjIsIm5iZiI6MTcxNzcyMDI2MiwicGF0aCI6Ii82Nzc3NzM2LzMzNzQ3ODQ3NS02NmMwZWU5Ni05NTllLTRjMGItODYxNC0xZTAzYTFjMTVmMDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDdUMDAzMTAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWVkMDY5ZWRlZDU5ODRkOTYwZWQ4NjAyYjZhYzgzMWY5OGFhMzM4ZmJkY2Q4NWY4ZTY2YTczNGQwZTlhNGE4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AeAwFL7PQ1IruQnG661acjA6qY6itY5zMXjr4iDvmmA">

Oh, that makes sense! I will add a clarifying note. It's easy to get confused about which version of the repo one is looking at.

clyne commented:
> > > > > > Also, a gh-pages website was never published, either in my personal repo or in the ProjectPythia repo for the cookbook. Not sure how to fix that.
> > > > > 
> > > > > 
> > > > > It's here: https://projectpythia.org/hello-pythia-cookbook/README.html
> > > > 
> > > > 
> > > > Sorry, should have been more clear. The gh-pages site is not linked from underneath the "About" heading on the repo [website](https://github.com/ProjectPythia/hello-pythia-cookbook). Shouldn't this happen automatically?
> > > 
> > > 
> > > No, it does not happen automatically. Maybe you missed my suggestion for this in [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book) under "Here's a handy trick..."
> > 
> > 
> > No, I didn't miss it, but it explains the rabbit hole that I went down. The "About" settings are different for my personal repo, and do not include the option to "Use your GitHub Pages website". I probably got confused about which repo I was looking at. Here's a screenshot for what I see on my personal repo:
> > <img alt="About" width="637" src="https://private-user-images.githubusercontent.com/6777736/337478475-66c0ee96-959e-4c0b-8614-1e03a1c15f05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3MjA1NjIsIm5iZiI6MTcxNzcyMDI2MiwicGF0aCI6Ii82Nzc3NzM2LzMzNzQ3ODQ3NS02NmMwZWU5Ni05NTllLTRjMGItODYxNC0xZTAzYTFjMTVmMDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDdUMDAzMTAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWVkMDY5ZWRlZDU5ODRkOTYwZWQ4NjAyYjZhYzgzMWY5OGFhMzM4ZmJkY2Q4NWY4ZTY2YTczNGQwZTlhNGE4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AeAwFL7PQ1IruQnG661acjA6qY6itY5zMXjr4iDvmmA">
> 
> Oh, that makes sense! I will add a clarifying note. It's easy to get confused about which version of the repo one is looking at.

Cool. I'll be going offline soon and have approved the PR. Nice work!

brian-rose commented on portal/cookbook-guide.md:
```diff
@@ -6,84 +6,257 @@ They are [geoscience](https://en.wikipedia.org/wiki/Earth_science)-focused
+- On the home page of your GitHub repo, look for the "About" section in the right-hand side bar.
+- Click the "gear" icon next to "About"
+- Select the checkbox "Use your GitHub Pages website".
+- Is that checkbox missing? That likely means you're looking at the "About" section of your _personal fork_, not the upstream fork. Navigate back to `https://github.com/ProjectPythia/[YOUR_COOKBOOK_REPO_NAME]/` and look for it there.
```
@clyne I added one last clarification here

brian-rose replied:
I should have merged the PR while I had approval, lol. If you have a chance to approve again, that would be terrific.

clyne replied:
Done!

brian-rose commented:
Thanks so much @clyne for the careful review! Much appreciated.
